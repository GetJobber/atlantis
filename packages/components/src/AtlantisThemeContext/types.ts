import { tokens } from "@jobber/design";
import { PropsWithChildren } from "react";

export interface AtlantisThemeContextValue {
  /**
   * The theme of the application.
   */
  readonly theme: Theme;

  /**
   * The design tokens for the current theme.
   */
  readonly tokens: typeof tokens;
}

export interface AtlantisThemeContextProviderProps extends PropsWithChildren {
  /**
   * The children to render.
   */
  readonly children: React.ReactNode;

  /**
   * Force the theme for this provider to always be the same as the provided theme. Useful for sections that should remain the same theme regardless of the rest of the application's theme.
   * This is dangerous because the children in this provider will not be able to change the theme.
   */
  readonly dangerouslyOverrideTheme?: Theme;
}

export type Theme = "light" | "dark";

export interface ThemeChangeDetails {
  theme: Theme;
}

export const THEME_CHANGE_EVENT = "atlantis.toggle-theme";
