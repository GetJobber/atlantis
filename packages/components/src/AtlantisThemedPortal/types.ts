import { PropsWithChildren } from "react";

export interface AtlantisThemedPortalProps extends PropsWithChildren {
  /** The DOM element to render the portal within. Defaults to `document.body`. */
  container?: HTMLElement;
}
