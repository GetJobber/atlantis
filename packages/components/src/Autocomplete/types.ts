import { XOR } from "ts-xor";
import { RefObject } from "react";
import { FormFieldProps } from "../FormField";

export type OptionCollection = XOR<Option[], GroupOption[]>;
export interface AutocompleteProps
  extends Pick<
    FormFieldProps,
    | "clearable"
    | "description"
    | "invalid"
    | "name"
    | "onBlur"
    | "onFocus"
    | "prefix"
    | "size"
    | "suffix"
    | "validations"
  > {
  /**
   * @deprecated
   * Use `ref` instead.
   */
  readonly inputRef?: FormFieldProps["inputRef"];

  /**
   * Initial options to show when user first focuses the Autocomplete
   */
  readonly initialOptions?: OptionCollection;

  /**
   * Set Autocomplete value.
   */
  readonly value: Option | undefined;

  /**
   * Allow the autocomplete to use values not from the drop down menu.
   *
   * @default true
   */
  readonly allowFreeForm?: boolean;

  /**
   * Debounce in milliseconds for getOptions
   *
   * @default 300
   */
  readonly debounce?: number;

  /**
   * Simplified onChange handler that only provides the new value.
   * @param newValue
   */
  onChange(newValue?: Option): void;

  /**
   * Called as the user types in the input. The autocomplete will display what
   * is returned from this method to the user as available options.
   * @param newInputText
   */
  getOptions(
    newInputText: string,
  ): OptionCollection | Promise<OptionCollection>;

  /**
   * Hint text that goes above the value once the form is filled out.
   */
  readonly placeholder: string;
}

export interface Option extends BaseOption {
  value?: OptionValue;
  description?: string;
  details?: string;
}

export interface GroupOption extends BaseOption {
  options: Option[];
}

export type AnyOption = XOR<Option, GroupOption>;

export type OptionValue = string | number;

export interface BaseOption {
  label: string;
}

export type UseAutocompleteProps = Pick<
  AutocompleteProps,
  "initialOptions" | "value" | "debounce" | "getOptions"
>;

export type UseAutocompleteHandlersProps = AutocompleteProps & {
  updateInput: (newText: string) => void;
  setMenuVisible: (visible: boolean) => void;
  inputText: string;
  setInputText: (newText: string) => void;
};

export enum IndexChange {
  Previous = -1,
  Next = 1,
}

export interface AutocompleteMenuProps {
  readonly visible: boolean;
  readonly options?: Option[];
  readonly selectedOption?: Option;
  /**
   * Element that it's attached to when the menu opens.
   */
  readonly attachTo: RefObject<Element | null>;
  onOptionSelect(chosenOption?: Option): void;
  setMenuRef(menuRef: HTMLElement | null): void;
  menuRef?: HTMLElement | null;
  attributes: Record<string, object | undefined>;
  addSeparators: boolean;
  targetWidth?: number;
  highlightedIndex: number;
  popperStyles: {
    [key: string]: React.CSSProperties;
  };
  isGroup(option?: AnyOption): boolean;
  isOptionSelected(
    selectedOption?: Option | undefined,
    option?: Option,
  ): boolean | undefined;
}

export interface UseKeyListenersProps {
  readonly options?: Option[];
  readonly visible: boolean;
  readonly menuRef?: HTMLElement | null;
  onOptionSelect(chosenOption?: Option): void;
  readonly highlightedIndex: number;
  readonly setHighlightedIndex: (index: number) => void;
}

export interface UseAutocompleteMenuProps {
  readonly options?: Option[];
  readonly visible: boolean;
  readonly menuRef?: HTMLElement | null;
  onOptionSelect(chosenOption?: Option): void;
}

export type AutoCompleteMenuOptionProps = Pick<
  AutocompleteMenuProps,
  | "highlightedIndex"
  | "addSeparators"
  | "isGroup"
  | "selectedOption"
  | "isOptionSelected"
  | "onOptionSelect"
> & { index: number; option: AutocompleteMenuProps["selectedOption"] };
