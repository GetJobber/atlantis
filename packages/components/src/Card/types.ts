import { ReactElement, ReactNode } from "react";
import colors from "./colors.css";
import { ButtonProps } from "../Button";
import { MenuProps } from "../Menu";

export type ActionProps = ReactElement<
  Omit<ButtonProps, "size" | "fullWidth"> | MenuProps
>;

export interface HeaderActionProps {
  /**
   * The title of the card.
   */
  readonly title?: string;
  /**
   * The action props that renders into a button on the card header.
   */
  readonly action?: ActionProps;
}

export interface CardProps {
  /**
   * The `accent`, if provided, will effect the color accent at the top of
   * the card.
   */
  readonly accent?: keyof typeof colors;
  readonly children: ReactNode | ReactNode[];
  /**
   * @deprecated
   * Use header instead.
   *
   */
  readonly title?: string;

  /**
   * The header props of the card.
   */
  readonly header?: string | HeaderActionProps | ReactElement;
}
