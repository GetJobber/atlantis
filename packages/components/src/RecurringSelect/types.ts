export enum WeekDay {
  Sunday,
  Monday,
  Tuesday,
  Wednesday,
  Thursday,
  Friday,
  Saturday,
}
export type DayOfMonth =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | "LAST";

export enum DurationPeriod {
  Day = "Day",
  Week = "Week",
  DayOfMonth = "DayOfMonth",
  WeekDayOfMonth = "WeekDayOfMonth",
  Year = "Year",
  Visit = "Visit",
}
export interface RecurrenceRuleDay {
  type: DurationPeriod.Day;
}

export interface RecurrenceRuleWeek {
  type: DurationPeriod.Week;
  weekDays: Set<WeekDay>;
}

export interface RecurrenceRuleYear {
  type: DurationPeriod.Year;
}

export interface RecurrenceRuleDayOfMonth {
  type: DurationPeriod.DayOfMonth;
  date: Set<DayOfMonth>;
}

export interface RecurrenceRuleWeekDayOfMonth {
  type: DurationPeriod.WeekDayOfMonth;
  dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
}

export type RecurrenceRule = {
  interval: number;
} & (
  | RecurrenceRuleDay
  | RecurrenceRuleWeek
  | RecurrenceRuleYear
  | RecurrenceRuleDayOfMonth
  | RecurrenceRuleWeekDayOfMonth
);
export interface InitialStateMap {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day;
  };
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week;
    weekDays: Set<WeekDay>;
  };
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth;
    date: Set<DayOfMonth>;
  };
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth;
    dayOfWeek: [Set<WeekDay>, Set<WeekDay>, Set<WeekDay>, Set<WeekDay>];
  };
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year;
  };
}

export const typeInitialStateMap: InitialStateMap = {
  [DurationPeriod.Day]: {
    type: DurationPeriod.Day,
  },
  [DurationPeriod.Week]: {
    type: DurationPeriod.Week,
    weekDays: new Set<WeekDay>(),
  },
  [DurationPeriod.DayOfMonth]: {
    type: DurationPeriod.DayOfMonth,
    date: new Set<DayOfMonth>(),
  },
  [DurationPeriod.WeekDayOfMonth]: {
    type: DurationPeriod.WeekDayOfMonth,
    dayOfWeek: [
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
      new Set<WeekDay>(),
    ],
  },
  [DurationPeriod.Year]: {
    type: DurationPeriod.Year,
  },
};

export const isMonthly = (
  rule: RecurrenceRule,
): rule is Extract<
  RecurrenceRule,
  { type: DurationPeriod.WeekDayOfMonth | DurationPeriod.DayOfMonth }
> =>
  rule.type === DurationPeriod.WeekDayOfMonth ||
  rule.type === DurationPeriod.DayOfMonth;
