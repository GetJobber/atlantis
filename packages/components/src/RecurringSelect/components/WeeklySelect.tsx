import React, { useId } from "react";
import styles from "./WeeklySelect.module.css";
import checkboxStyles from "../DateCellCheckbox.module.css";
import { WeekDay } from "../types";

/**
 * @deprecated WeeklySelect will be removed in the next major version
 */
interface WeeklySelectProps {
  readonly disabled: boolean;
  readonly selectedDays: Set<WeekDay>;
  onChange(selectedDays: Set<WeekDay>): void;
}
const weekDays = [0, 1, 2, 3, 4, 5, 6];

/**
 * @deprecated WeeklySelect will be removed in the next major version
 */
export function WeeklySelect({
  onChange,
  selectedDays,
  disabled,
}: WeeklySelectProps) {
  const onChangeWeekDays = (event: React.FormEvent<HTMLInputElement>): void => {
    const day = Number(event.currentTarget.value);
    const next = new Set(Array.from(selectedDays));

    if (selectedDays.has(day)) {
      next.delete(day);
    } else {
      next.add(day);
    }
    onChange(next);
  };

  return (
    <div className={styles.container}>
      {weekDays.map(weekDay => {
        const isSelected = selectedDays.has(weekDay);
        const inputId = useId();

        return (
          <div key={`${weekDay}`} className={checkboxStyles.checkboxWrapper}>
            <input
              className={checkboxStyles.dayCheckboxInput}
              disabled={disabled}
              type="checkbox"
              onChange={onChangeWeekDays}
              id={`weekly-select-${weekDay}${inputId}`}
              aria-checked={isSelected}
              checked={isSelected}
              value={weekDay}
            />
            <label
              className={checkboxStyles.dayCheckbox}
              htmlFor={`weekly-select-${weekDay}${inputId}`}
            >
              {`${WeekDay[weekDay].substr(0, 1).toUpperCase()}`}
            </label>
          </div>
        );
      })}
    </div>
  );
}
