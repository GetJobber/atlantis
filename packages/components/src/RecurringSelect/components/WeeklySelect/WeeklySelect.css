/* 
  faking a border with a background on the parent element, gaps & padding
  to prevent double borders on overlapping boxes
*/
.container {
  display: grid;
  width: 100%;
  padding: var(--space-smaller);
  border: var(--border-base) solid var(--color-border);
  border-radius: var(--radius-large);
  grid-template-columns: repeat(7, 1fr);
  row-gap: var(--space-smaller);
  column-gap: var(--space-smaller);
}

.checkboxWrapper {
  position: relative;
}

/* kept present but hidden for accessibility */
.container input[type="checkbox"] {
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(50%);
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  overflow: hidden;
}

.dayCheckbox {
  display: block;
  padding: 1rem;
  border-radius: var(--radius-large);
  text-align: center;
  user-select: none;
  background: var(--color-surface);
  transition: 300ms;
}

.dayCheckbox:hover {
  background: var(--color-surface--hover);
}

.container input[type="checkbox"]:focus + .dayCheckbox {
  box-shadow: var(--shadow-focus);
  background: var(--color-surface--hover);
}

.container input[type="checkbox"]:checked + .dayCheckbox {
  color: var(--color-white);
  background: var(--color-interactive);
}

.container input[type="checkbox"]:disabled + .dayCheckbox {
  color: var(--color-disabled);
  background: var(--color-surface--background);
  cursor: default;
}

.container input[type="checkbox"]:enabled + .dayCheckbox {
  cursor: pointer;
}

.container input[type="checkbox"]:checked:focus + .dayCheckbox {
  background: var(--color-interactive--hover);
}

.container input[type="checkbox"]:checked + .dayCheckbox:hover {
  background: var(--color-interactive--hover);
}

.container input[type="checkbox"]:disabled:checked + .dayCheckbox {
  color: var(--color-disabled--secondary);
  background: var(--color-disabled);
}
