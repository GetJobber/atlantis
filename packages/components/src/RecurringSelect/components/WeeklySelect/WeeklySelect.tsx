import React from "react";
import styles from "./WeeklySelect.css";
import checkboxStyles from "../../DateCellCheckbox.css";
import { WeekDay } from "../../types";

interface WeeklySelectProps {
  disabled: boolean;
  disambiguator?: string;
  selectedDays: Set<WeekDay>;
  onChange(selectedDays: Set<WeekDay>): void;
}
const weekDays = [0, 1, 2, 3, 4, 5, 6];

export function WeeklySelect({
  onChange,
  disambiguator = "",
  selectedDays,
  disabled,
}: WeeklySelectProps) {
  const onChangeWeekDays = (event: React.FormEvent<HTMLInputElement>): void => {
    const day = Number(event.currentTarget.value);
    const next = new Set(Array.from(selectedDays));

    if (selectedDays.has(day)) {
      next.delete(day);
    } else {
      next.add(day);
    }
    onChange(next);
  };

  return (
    <div className={styles.container}>
      {weekDays.map(weekDay => {
        const isSelected = selectedDays.has(weekDay);

        return (
          <div key={`${weekDay}`} className={checkboxStyles.checkboxWrapper}>
            <input
              className={checkboxStyles.dayCheckboxInput}
              disabled={disabled}
              type="checkbox"
              onChange={onChangeWeekDays}
              id={`weekly-select-${weekDay}${disambiguator}`}
              aria-checked={isSelected}
              checked={isSelected}
              value={weekDay}
            />
            <label
              className={checkboxStyles.dayCheckbox}
              htmlFor={`weekly-select-${weekDay}${disambiguator}`}
            >
              {`${WeekDay[weekDay].substr(0, 1).toUpperCase()}`}
            </label>
          </div>
        );
      })}
    </div>
  );
}
