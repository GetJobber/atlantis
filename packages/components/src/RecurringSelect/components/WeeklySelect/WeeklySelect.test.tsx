/* tslint:disable no-relative-imports */
import React from "react";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { WeeklySelect } from "./WeeklySelect";
// eslint-disable-next-line
import { DurationPeriod, WeekDay } from "../../../Scheduler/types";

afterEach(cleanup);

let onChange: jest.Mock;

describe("WeeklySelect", () => {
  beforeEach(() => {
    onChange = jest.fn();
  });

  it("should render the component", () => {
    const tree = renderer
      .create(
        <WeeklySelect
          disabled={false}
          onChange={onChange}
          selectedDays={new Set<WeekDay>()}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
