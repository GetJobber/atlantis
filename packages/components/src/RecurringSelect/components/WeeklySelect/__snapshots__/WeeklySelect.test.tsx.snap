// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeeklySelect should render the component 1`] = `
<div
  className="container"
>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-0"
      onChange={[Function]}
      type="checkbox"
      value={0}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-0"
    >
      S
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-1"
      onChange={[Function]}
      type="checkbox"
      value={1}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-1"
    >
      M
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-2"
      onChange={[Function]}
      type="checkbox"
      value={2}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-2"
    >
      T
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-3"
      onChange={[Function]}
      type="checkbox"
      value={3}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-3"
    >
      W
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-4"
      onChange={[Function]}
      type="checkbox"
      value={4}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-4"
    >
      T
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-5"
      onChange={[Function]}
      type="checkbox"
      value={5}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-5"
    >
      F
    </label>
  </div>
  <div
    className="checkboxWrapper"
  >
    <input
      aria-checked={false}
      checked={false}
      className="dayCheckboxInput"
      disabled={false}
      id="weekly-select-6"
      onChange={[Function]}
      type="checkbox"
      value={6}
    />
    <label
      className="dayCheckbox"
      htmlFor="weekly-select-6"
    >
      S
    </label>
  </div>
</div>
`;
